src/exceptions.jl

```diff
@@ -0,0 +1,101 @@
+"Base abstract type for all custom exceptions thrown by LibPQ.jl"
+abstract type LibPQException <: Exception end
+
+"An exception with an error message generated by PostgreSQL"
+abstract type PostgreSQLException <: LibPQException end
+
+# PostgreSQL errors have trailing newlines
+# https://www.postgresql.org/docs/10/libpq-status.html#LIBPQ-PQERRORMESSAGE
+Base.showerror(io::IO, err::PostgreSQLException) = print(io, chomp(err.msg))
+
+"An exception generated by LibPQ.jl"
+abstract type JLClientException <: LibPQException end
+
+"An error regarding a connection reported by PostgreSQL"
+struct PQConnectionError <: PostgreSQLException
+    msg::String
+end
+
+function PQConnectionError(jl_conn::Connection)
+    return PQConnectionError(error_message(jl_conn))
+end
+
+"An error regarding a connection reported by PostgreSQL"
+struct ConninfoParseError <: PostgreSQLException
+    msg::String
+end
+
+"An error regarding a connection generated by LibPQ.jl"
+struct JLConnectionError <: JLClientException
+    msg::String
+end
+
+"An error regarding a query result generated by LibPQ.jl"
+struct JLResultError <: JLClientException
+    msg::String
+end
+
+"""
+An error regarding a query result generated by PostgreSQL
+
+The `Code` parameter represents the PostgreSQL error code as defined in
+[Appendix A. PostgreSQL Error Codes](https://www.postgresql.org/docs/devel/errcodes-appendix.html).
+The `Class` parameter is the first two characters of that code, also listed on that page.
+
+For a list of all error aliases, see [src/error_codes.jl](https://github.com/invenia/LibPQ.jl/blob/master/src/error_codes.jl),
+which was generated using the PostgreSQL documentation linked above.
+
+```jldoctest
+julia> try execute(conn, "SELORCT NUUL;") catch err println(err) end
+LibPQ.SyntaxError("ERROR:  syntax error at or near \\"SELORCT\\"\\nLINE 1: SELORCT NUUL;\\n        ^\\n")
+
+julia> LibPQ.SyntaxError
+LibPQ.PQResultError{c"42",e"42601"}
+```
+"""
+struct PQResultError{Class, Code} <: PostgreSQLException
+    msg::String
+    verbose_msg::Union{String, Nothing}
+
+    function PQResultError{Class_, Code_}(msg, verbose_msg) where {Class_, Code_}
+        return new{Class_::Class, Code_::ErrorCode}(
+            convert(String, msg),
+            convert(Union{String, Nothing}, verbose_msg),
+        )
+    end
+end
+
+include("error_codes.jl")
```

Comment:
related: why not put them in there own submodule?
```suggestion
module Errors
    include("error_codes.jl")
end
```